### ISSUE TITLE
Basic Auth from uri

### ISSUE THREAD

**Issue by dalefrancis88** [IdentityServer/IdentityServer3#3272]
- [x] I read and understood how to enable [logging](https://identityserver.github.io/Documentation/docsv2/configuration/logging.html)

### Question / Issue
Guys,

I have a requirment to add authentication to the webhooks in our solution. 

When we make a call to a third-party we specify the return url. On here I can add the username and password to the callback url in the form http://username:password@domain.com/callback?querystring=here

I have looked around the notes and can't see any obvious way to implement this with IdentityServer3. So i wanted to reach out and see if you could let me know if i've just missed something in the docs or if my idea is workable.

1) Put the client id and a long lived token as the username and password, extract it using middleware and set the Auth header

2) Implement the Thinktecture.IdentityModel.Owin.BasicAuthentication and have the webapi pass engine:token and call ValidateAccessToken, extract the claims and return them

3) Implement the BasicAuth, create a short lifetime client per third party using the client credentials flow, have the url have clientid:secret then call the token endpoint to sign in.

4) open to ideas?

**Comment by leastprivilege** [comment]
No idea - have you tried it?

**Comment by dalefrancis88** [comment]
Yeah I came up with a solution that works for now. 

I created an authentication middleware. OWIN asks the middle ware to authenticate, when that fails OWIN  calls the ApplyResponceChallengeAsync method. This challenge will cause the Basic auth creds to be stripped from the URL and added as an Authorization header.

        protected override Task ApplyResponseChallengeAsync()
        {
            if (Response.StatusCode == 401 && Helper.LookupChallenge(Options.AuthenticationType, Options.AuthenticationMode) != null)
                Response.Headers.AppendValues("WWW-Authenticate", _challenge);
            return Task.FromResult((object)null);
        }

From here it looks at the auth header to see if something exists and to double check that its a BASIC token. I then strip out the username(client) and password(token) by decoding from Base64 and i switch out the BASIC token for a BEARER token to be later verified by IdentityServer
